hasktorch/src/Torch/Vision/Darknet/Forward.hs

```diff
@@ -0,0 +1,491 @@
+  sample S.ShortCutSpec {..} = do
+    ShortCut
+      <$> pure from
+
+instance HasForward ShortCut (Map Int Tensor) Tensor where
+  forward ShortCut {..} inputs = inputs ! from
+
+type Anchors = [(Float,Float)]
+type ScaledAnchors = [(Float,Float)]
+
+data Yolo
+  = Yolo
+      { anchors :: Anchors,
+        classes :: Int,
+        img_size :: Int
+      }
+  deriving (Show, Generic, Parameterized)
+
+instance Randomizable S.YoloSpec Yolo where
+  sample S.YoloSpec {..} = pure $ Yolo { classes = classes ,
+                                         anchors = map (\(a,b) -> (fromIntegral a,fromIntegral b))anchors,
+                                         img_size = img_size
+                                       
+                                       }
+
+newtype Prediction = Prediction {fromPrediction :: Tensor} deriving (Show)
+
+toPrediction :: Yolo -> Tensor -> Prediction
+toPrediction Yolo {..} input =
+  let num_samples = D.size input 0
+      grid_size = D.size input 2
+      num_anchors = length anchors
+   in Prediction $ D.contiguous $ D.permute [0, 1, 3, 4, 2] $ D.view [num_samples, num_anchors, classes + 5, grid_size, grid_size] input
+
+squeezeLastDim :: Tensor -> Tensor
+squeezeLastDim input = I.squeezeDim input (-1)
+
+toX ::
+  -- | [batch, anchors, grid, grid, class + 5]
+  Prediction ->
+  -- |  [batch, anchors, grid, grid]
+  Tensor
+toX prediction = D.sigmoid $ squeezeLastDim (D.slice (-1) 0 1 1 $ fromPrediction prediction)
+
+toY ::
+  -- | [batch, anchors, grid, grid, class + 5]
+  Prediction ->
+  -- |  [batch, anchors, grid, grid]
+  Tensor
+toY prediction = D.sigmoid $ squeezeLastDim (D.slice (-1) 1 2 1 $ fromPrediction prediction)
+
+toW ::
+  -- | [batch, anchors, grid, grid, class + 5]
+  Prediction ->
+  -- |  [batch, anchors, grid, grid]
+  Tensor
+toW prediction = squeezeLastDim (D.slice (-1) 2 3 1 $ fromPrediction prediction)
+
+toH ::
+  -- | [batch, anchors, grid, grid, class + 5]
+  Prediction ->
+  -- |  [batch, anchors, grid, grid]
+  Tensor
+toH prediction = squeezeLastDim (D.slice (-1) 3 4 1 $ fromPrediction prediction)
+
+toPredConf ::
+  -- | [batch, anchors, grid, grid, class + 5]
+  Prediction ->
+  -- |  [batch, anchors, grid, grid]
+  Tensor
+toPredConf prediction = D.sigmoid $ squeezeLastDim (D.slice (-1) 4 5 1 $ fromPrediction prediction)
+
+toPredClass ::
+  -- | [batch, anchors, grid, grid, class + 5]
+  Prediction ->
+  -- |  [batch, anchors, grid, grid, class]
+  Tensor
+toPredClass prediction =
+  let input = fromPrediction prediction
+      num_features = D.size input (-1)
+   in D.sigmoid $ squeezeLastDim (D.slice (-1) 5 num_features 1 input)
+
+gridX ::
+  -- |  grid size
+  Int ->
+  -- |  [1, 1, grid, grid]
+  Tensor
+gridX g = D.view [1, 1, g, g] $ D.repeat [g, 1] $ arange' (0 :: Int) g (1 :: Int)
+
+gridY ::
+  -- |  grid size
+  Int ->
+  -- |  [1, 1, grid, grid]
+  Tensor
+gridY g = D.contiguous $ D.view [1, 1, g, g] $ I.t $ D.repeat [g, 1] $ arange' (0 :: Int) g (1 :: Int)
```

Comment:
https://stackoverflow.com/questions/49643225/whats-the-difference-between-reshape-and-view-in-pytorch
D.contiguous $ D.view  is the same as D.reshape